domain/secret/state/state.go

```diff
@@ -1687,6 +1707,184 @@ ON CONFLICT(secret_id, unit_uuid) DO UPDATE SET
+			return errors.Trace(err)
+		} else if !isLocal {
+			// Should never happen.
+			return secreterrors.SecretIsNotLocal
+		}
+
+		err = tx.Query(ctx, queryStmt, consumer).GetAll(&dbSecretConsumers)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotate(err, "querying secret consumers")
+		}
+
+		result := secretInfo{ID: uri.ID}
+		err = tx.Query(ctx, selectLatestRevisionStmt, result).Get(&result)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return secreterrors.SecretNotFound
+			} else {
+				return errors.Annotatef(err, "looking up latest revision for %q", uri.ID)
+			}
+		}
+		latestRevision = result.LatestRevision
+
+		return nil
+	})
+	if err != nil {
+		return nil, 0, errors.Trace(domain.CoerceError(err))
+	}
+	if len(dbSecretConsumers) == 0 {
+		return nil, latestRevision, fmt.Errorf("secret consumer for %q and unit %q%w", uri.ID, unitName, secreterrors.SecretConsumerNotFound)
+	}
+	consumers, err := dbSecretConsumers.toSecretConsumers()
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+	return consumers[0], latestRevision, nil
+}
+
+// SaveSecretRemoteConsumer saves the consumer metadata for the given secret and unit.
+// If the secret does not exist, an error satisfying [secreterrors.SecretNotFound] is returned.
+func (st State) SaveSecretRemoteConsumer(ctx context.Context, uri *coresecrets.URI, unitName string, md *coresecrets.SecretConsumerMetadata) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	insertQuery := `
+INSERT INTO secret_remote_unit_consumer (*)
+VALUES ($secretRemoteUnitConsumer.*)
+ON CONFLICT(secret_id, unit_id) DO UPDATE SET
+    current_revision=excluded.current_revision
+`
+
+	insertStmt, err := st.Prepare(insertQuery, secretRemoteUnitConsumer{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	consumer := secretRemoteUnitConsumer{
+		SecretID:        uri.ID,
+		UnitID:          unitName,
+		CurrentRevision: md.CurrentRevision,
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if isLocal, err := st.checkExistsIfLocal(ctx, tx, uri); err != nil {
+			return errors.Trace(err)
+		} else if !isLocal {
+			// Should never happen.
+			return secreterrors.SecretIsNotLocal
+		}
+		if err := tx.Query(ctx, insertStmt, consumer).Run(); err != nil {
+			return errors.Trace(err)
+		}
+
+		if err := st.markObsoleteRevisions(ctx, tx, uri); err != nil {
+			return errors.Annotatef(err, "marking obsolete revisions for secret %q", uri)
+		}
+
+		return nil
+	})
+	return errors.Trace(domain.CoerceError(err))
+}
+
+// UpdateRemoteSecretRevision records the latest revision of the specified cross model secret.
+func (st State) UpdateRemoteSecretRevision(ctx context.Context, uri *coresecrets.URI, latestRevision int) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	insertQuery := `
+INSERT INTO secret (id)
+VALUES ($secretID.id)
+ON CONFLICT(id) DO NOTHING
```

Comment:
Do we need to do this?
Probably change the secret_id in secret_reference to non-FK.

Reply:
Yes because this is a single column table so if the secret id exists already when we upsert, there's nothing to do.
We don't want to compromise RI by removing FK.
This table is similar to the net_node table.